Refactor Symint Deduping to separate pass

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118771
* __->__ #118938

Previously Symint Deduping was done during proxy tracing which made it more difficult to reason about. This refactors the deduping to a separate pass. 

We only dedupe symints which are resolvable from input symint nodes so as to avoid inducing a dependency on the backward in the forward.

potential fix for : https://github.com/pytorch/pytorch/issues/118224


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

eellison committed:
Refactor Symint Deduping to separate pass

[ghstack-poisoned]

eellison committed:
Update on "Refactor Symint Deduping to separate pass"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Refactor Symint Deduping to separate pass"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Refactor Symint Deduping to separate pass"


Previously Symint Deduping was done during proxy tracing which made it more difficult to reason about. This refactors the deduping to a separate pass. 

We only dedupe symints which are resolvable from input symint nodes so as to avoid inducing a dependency on the backward in the forward.

potential fix for : https://github.com/pytorch/pytorch/issues/118224


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Refactor Symint Deduping to separate pass"


Previously Symint Deduping was done during proxy tracing which made it more difficult to reason about. This refactors the deduping to a separate pass. 

We only dedupe symints which are resolvable from input symint nodes so as to avoid inducing a dependency on the backward in the forward.

potential fix for : https://github.com/pytorch/pytorch/issues/118224


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -633,6 +548,8 @@ def dispatch_trace(
 ) -> GraphModule:
     graph = tracer.trace(root, concrete_args)
+    from torch._inductor.fx_passes.dedupe_symint_uses import dedupe_symints
+    dedupe_symints(graph)
```
oh you want to unconditionally call it here? I guess we can...

eellison replied:
Right, that was the previous state

ezyang commented on torch/_inductor/fx_passes/dedupe_symint_uses.py:
```diff
@@ -0,0 +1,78 @@
+    def __eq__(self, value) -> bool:
+        if not isinstance(value, _SymExprHash):
+            return False
+        return self.sym_obj.node.expr == value.sym_obj.node.expr
```
Not going to test for type equality?

ezyang commented on torch/_inductor/fx_passes/dedupe_symint_uses.py:
```diff
@@ -0,0 +1,78 @@
+    """
+
+    sym_dict = _SymHashingDict()
+    resolvable_from_input_symints = set()
```
Aren't things that are resolvable from inputs guaranteed to be symbols, rather than expressions? (I suppose in a rare occasion we could have created an input for a variable s0 which later got replaced into s1 + 3, but in this case it doesn't seem like a big deal to fail to dedupe to this input...)

ezyang commented on torch/_inductor/fx_passes/dedupe_symint_uses.py:
```diff
@@ -0,0 +1,78 @@
+            graph.erase_node(node)
+        elif all(n in resolvable_from_input_symints for n in node.all_input_nodes):
+            sym_dict[val] = node
+            resolvable_from_input_symints.add(node)
```
Except this seems to imply that you're propagating this recursively? So what is even the point--the algorithm right now seems to only exclude nodes from being resolved if the SymInt has never been seen before in the graph.

ezyang commented on torch/_inductor/fx_passes/dedupe_symint_uses.py:
```diff
@@ -0,0 +1,78 @@
+            sym_dict[val] = node
+        elif existing_node := sym_dict.get(val):
+            node.replace_all_uses_with(existing_node)
+            graph.erase_node(node)
```
So I am concerned that you may need to replay the computation, and just looking for the SymInt may not be enough. Consider the following graph:

```
def f(x):
  y = torch.cat((x, x))
  z = y.size(0)
  ... something involving z ...
```

Given that x: f32[s0], the val for z is `s0 * 2`. This expression has never appeared, except from y. So you will fail to actually generate a size expression that doesn't keep a tensor live, since you're going to hang it off of `torch.cat`. Actually, I guess you are calling this a SymInt deduping pass, so I suppose it isn't obvious the pass is supposed to deal with this, but I could swear part of the reason you were doing was to eliminate false dependencies on tensors that weren't actually real?

ezyang replied:
Calling back from Jun 2024, I need to do this version of the pass :P

ezyang reviewed (approved):
I'm approving because this is a strict improvement over the existing situation, but I do have questions on the algo.

eellison committed:
Update on "Refactor Symint Deduping to separate pass"


Previously Symint Deduping was done during proxy tracing which made it more difficult to reason about. This refactors the deduping to a separate pass. 

We only dedupe symints which are resolvable from input symint nodes so as to avoid inducing a dependency on the backward in the forward.

potential fix for : https://github.com/pytorch/pytorch/issues/118224


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
